tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_corrective.ksh

```diff
@@ -0,0 +1,135 @@
+#!/bin/ksh -p
+#
+# CDDL HEADER START
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2019 Datto, Inc. All rights reserved.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+# ZFS be able to heal using corrective recv
+#
+# STRATEGY:
+#  1. Create a dataset
+#  2. Snapshot the dataset
+#  3. Create a file and get its checksum
+#  4. Snapshot the dataset
+#  5. Recv dataset into a filesystem with different compression
+#  5. Corrupt the file
+#  6. Heal the corruption using a corrective send and full send file
+#  7. Corrupt the file again
+#  8. Heal the corruption using a corrective send an incremental send file
+#  9. Corrupt the file again
+# 10. Heal the corruption when the target snapshot and the send file have
+#     different compressions algorithms
+#
+
+verify_runnable "both"
+
+backup=$TEST_BASE_DIR/backup
+ibackup=$TEST_BASE_DIR/ibackup.$$
+garbage=$TEST_BASE_DIR/garbage.$$
+DISK=${DISKS%% *}
+
+function cleanup
+{
+	datasetexists $TESTPOOL/$TESTFS1 && \
+		log_must zfs destroy -r $TESTPOOL/$TESTFS1
+	datasetexists $TESTPOOL/$TESTFS2 && \
+		log_must zfs destroy -r $TESTPOOL/$TESTFS2
+
+	for file in $garbage $ibackup $backup; do
+		[[ -f $file ]] && log_must rm -f $file
+	done
+}
+
+function corrupt_offset
+{
+	log_must dd bs=512 count=1 if=garbage conv=notrunc \
+	    oflag=sync of=$DEV_RDSKDIR/$DISK seek=$((($1 / 512) + (0x400000 / 512)))
+}
+
+function test_corrective_recv
+{
+	typeset objid="$(zdb -dd $1 | sed -n '/file/s/ \+/ /gp' | \
+	    cut -f2 -d' ')"
+	typeset start_offset="0x$(zdb -ddddd $1 $objid | \
+	    sed -n '/ 0   L0 0/s/ \+/ /gp' | cut -f2 -d':')"
+	typeset mid_offset="0x$(zdb -ddddd $1 $objid | \
+	    sed -n '/ 27e00   L0 0:/s/ \+/ /gp' | cut -f2 -d':')"
+
+	corrupt_offset "$start_offset"
+	corrupt_offset "$mid_offset"
+
+	log_must zpool scrub $TESTPOOL
+	log_must sleep 5 # let scrub finish
+	log_must eval "zpool status -v $TESTPOOL | \
+	    grep \"Permanent errors have been detected\""
+
+	# make sure we will read the corruption from disk by flushing the ARC
+	log_must zinject -a
+
+	log_must eval "zfs recv -c $1 < $2"
+
+	log_mustnot eval "zpool status -v $TESTPOOL | \
+	    grep \"Permanent errors have been detected\""
+	typeset cksum=$(md5digest $file)
+	[[ "$cksum" == "$checksum" ]] || \
+		log_fail "Checksums differ ($cksum1 != $checksum)"
+}
+
+log_onexit cleanup
+
+log_assert "ZFS corrective receive should be able to heal corruption"
+
+# we will use this as the source of corruption
+log_must dd if=/dev/urandom of=garbage bs=512 count=1 oflag=sync
```

Comment:
You should be able to get rid of the garbage file after switching to the targeted injection.